diff --git a/nova/virt/xenapi/vmops.py b/nova/virt/xenapi/vmops.py
index 82a9aef..f65b7de 100644
--- a/nova/virt/xenapi/vmops.py
+++ b/nova/virt/xenapi/vmops.py
@@ -2278,10 +2278,11 @@ class VMOps(object):
                 self._call_live_migrate_command(
                     "VM.assert_can_migrate", vm_ref, dest_check_data)
             except self._session.XenAPI.Failure as exc:
-                reason = exc.details[0]
-                msg = _('assert_can_migrate failed because: %s') % reason
-                LOG.debug(msg, exc_info=True)
-                raise exception.MigrationPreCheckError(reason=msg)
+                reason = '%s' % exc.details[0]
+                if reason.strip().upper() != "VIF_NOT_IN_MAP":
+                    msg = _('assert_can_migrate failed because: %s') % reason
+                    LOG.debug(msg, exc_info=True)
+                    raise exception.MigrationPreCheckError(reason=msg)
         return dest_check_data
 
     def _ensure_pv_driver_info_for_live_migration(self, instance, vm_ref):
