diff --git a/ceilometer/compute/virt/xenapi/inspector.py b/ceilometer/compute/virt/xenapi/inspector.py
index d5676b1..6048e3a 100644
--- a/ceilometer/compute/virt/xenapi/inspector.py
+++ b/ceilometer/compute/virt/xenapi/inspector.py
@@ -238,10 +238,3 @@ class XenapiInspector(virt_inspector.Inspector):
                     write_bytes_rate=write_rate,
                     write_requests_rate=0)
                 yield(disk, disk_rate_info)
-
-    def purge_inspection_cache(self):
-        # Empty function to fit MOS9.2 where get_samples will invoke
-        # self.inspector.purge_inspection_cache(); will envalue if
-        # support cache for XenAPI. But obviously we need the
-        # the code for caching available in upstream firstly.
-        pass
